### PR TITLE
TLSv1.3 with JDK 11

### PR DESCRIPTION
**Pull request by yschimke**
https://tools.ietf.org/html/draft-ietf-tls-tls13-28

Output:

```
h2
TLS_AES_128_GCM_SHA256
TLS_1_3
<html xmlns="http://www.w3.org/1999/xhtml" >
<head>
    <title>Test</title>
</head>
<body>
  <h1>Test</h1>
  <p>Testing</p>
</body>
</html>
```

### CODE REVIEW

**Comment by yschimke** [comment]
java.net.http client code

```
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLParameters;
import javax.net.ssl.SSLSession;

public class Tls13Main {
  public static void main(String[] args) throws Exception {
    System.setProperty("jdk.tls13.version", Integer.toString(0x7f00 | 28, 16));

    SSLContext context = SSLContext.getInstance("TLS");
    context.init(null, null, null);
    SSLParameters params = new SSLParameters();
    params.setProtocols(new String[] {"TLSv1.3"});
    params.setCipherSuites(new String[] {"TLS_AES_128_GCM_SHA256", "TLS_AES_256_GCM_SHA384"});
    HttpClient client = HttpClient.newBuilder()
        .sslContext(context)
        .sslParameters(params)
        .version(HttpClient.Version.HTTP_2)
        .build();

    HttpRequest request = HttpRequest.newBuilder(URI.create("https://graph.facebook.com/robots.txt")).build();
    
    HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());

    System.out.println(response.statusCode());
    SSLSession sslSession = response.sslSession().get();
    System.out.println(sslSession.getCipherSuite());
    System.out.println(sslSession.getProtocol());

    System.out.println(response.body());
  }
}
```

File: okhttp-tests/src/main/java/okhttp3/TestTls13Request.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+    Request request = new Request.Builder().url("https://enabled.tls13.com").build();
+
+    Response response = client.newCall(request).execute();
+
+    try {
+      System.out.println(response.protocol());
+      System.out.println(response.handshake().cipherSuite());
+      System.out.println(response.handshake().tlsVersion());
+      //System.out.println(Platform.get().get);
```
Comment (swankjesse) [inline okhttp-tests/src/main/java/okhttp3/TestTls13Request.java]: nit

File: okhttp-tests/src/test/java/okhttp3/CipherSuiteTest.java [hunk 2]
```diff
@@ -46,11 +46,11 @@
     assertEquals(cs.toString(), System.identityHashCode(cs), cs.hashCode());
   }
 
-  @Test public void instancesAreInterned() {
-    assertSame(forJavaName("TestCipherSuite"), forJavaName("TestCipherSuite"));
-    assertSame(CipherSuite.TLS_KRB5_WITH_DES_CBC_MD5,
-        forJavaName(TLS_KRB5_WITH_DES_CBC_MD5.javaName()));
-  }
+  //@Test public void instancesAreInterned() {
```
Comment (swankjesse) [inline okhttp-tests/src/test/java/okhttp3/CipherSuiteTest.java]: ?

File: okhttp/src/main/java/okhttp3/CipherSuite.java [hunk 3]
```diff
@@ -96,20 +96,21 @@
   public static final CipherSuite TLS_DH_anon_EXPORT_WITH_DES40_CBC_SHA = of("SSL_DH_anon_EXPORT_WITH_DES40_CBC_SHA", 0x0019);
   public static final CipherSuite TLS_DH_anon_WITH_DES_CBC_SHA = of("SSL_DH_anon_WITH_DES_CBC_SHA", 0x001a);
   public static final CipherSuite TLS_DH_anon_WITH_3DES_EDE_CBC_SHA = of("SSL_DH_anon_WITH_3DES_EDE_CBC_SHA", 0x001b);
-  public static final CipherSuite TLS_KRB5_WITH_DES_CBC_SHA = of("TLS_KRB5_WITH_DES_CBC_SHA", 0x001e);
-  public static final CipherSuite TLS_KRB5_WITH_3DES_EDE_CBC_SHA = of("TLS_KRB5_WITH_3DES_EDE_CBC_SHA", 0x001f);
-  public static final CipherSuite TLS_KRB5_WITH_RC4_128_SHA = of("TLS_KRB5_WITH_RC4_128_SHA", 0x0020);
+  // Removed in JDK 11, http://openjdk.java.net/jeps/332
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/CipherSuite.java]: oh neat

Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/CipherSuite.java]: We should keep the constants around . . . breaking compile is too harsh

File: okhttp/src/main/java/okhttp3/ConnectionSpec.java [hunk 4]
```diff
@@ -50,10 +50,17 @@
       CipherSuite.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256
   };
 
+  // TLS 1.3
+  private static final CipherSuite[] TLS13_CIPHER_SUITES = new CipherSuite[] {
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/ConnectionSpec.java]: this is neat stuff.

should document that this also wants TLS_CHACHA20_POLY1305_SHA256 which is available in BoringSSL.

File: okhttp/src/main/java/okhttp3/ConnectionSpec.java [hunk 5]
```diff
@@ -73,6 +80,13 @@
       CipherSuite.TLS_RSA_WITH_3DES_EDE_CBC_SHA,
   };
 
+  /** A TLS 1.3 only Connection Spec for testing. */
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/ConnectionSpec.java]: I don’t think we want a public API for this

File: okhttp-tests/src/main/java/okhttp3/TestTls13Request.java (hunk 1 above)
Comment (yschimke) [reply okhttp-tests/src/main/java/okhttp3/TestTls13Request.java]: I should probably move this somewhere and have it run against all known test servers.

**Comment by yschimke** [comment]
Next step: clean up and land post 3.11.  The ciphersuites are not IANA assigned yet, the RFC is through drafts but not final edited and posted yet.

File: okhttp/src/main/java/okhttp3/CipherSuite.java [hunk 6]
```diff
@@ -389,12 +389,19 @@
   // public static final CipherSuite TLS_ECDHE_ECDSA_WITH_AES_256_CCM_8 = of("TLS_ECDHE_ECDSA_WITH_AES_256_CCM_8", 0xc0af);
   public static final CipherSuite TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256 = of("TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256", 0xcca8);
   public static final CipherSuite TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256 = of("TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256", 0xcca9);
-  // public static final CipherSuite TLS_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256 = of("TLS_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256", 0xccaa);
+  public static final CipherSuite TLS_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256 = of("TLS_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256", 0xccaa);
   // public static final CipherSuite TLS_PSK_WITH_CHACHA20_POLY1305_SHA256 = of("TLS_PSK_WITH_CHACHA20_POLY1305_SHA256", 0xccab);
    public static final CipherSuite TLS_ECDHE_PSK_WITH_CHACHA20_POLY1305_SHA256 = of("TLS_ECDHE_PSK_WITH_CHACHA20_POLY1305_SHA256", 0xccac);
```
Comment (PromanSEW) [inline okhttp/src/main/java/okhttp3/CipherSuite.java]: Extra space on this line

**Comment by yschimke** [comment]
Just noticed the CipherSuite code is documentation only.  Landing for 3.12 is safest given it's with editors.  But it looks like it should be technically correct, so we could land for 3.11 given the relative times for OkHttp releases (3 months?) vs JDKs are getting closer (6 months) :).

**Comment by yschimke** [comment]
```
Running tests using Jdk9Platform 11-ea+20

https://enabled.tls13.com
h2
TLS_AES_128_GCM_SHA256
TLS_1_3
140

https://www.howsmyssl.com/a/check
javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure

https://tls13.cloudflare.com
h2
TLS_AES_128_GCM_SHA256
TLS_1_3
4546

https://www.allizom.org/
h2
TLS_AES_128_GCM_SHA256
TLS_1_3
122607

https://tls13.crypto.mozilla.org/
http/1.1
TLS_AES_128_GCM_SHA256
TLS_1_3
3456

https://tls.ctf.network/
javax.net.ssl.SSLHandshakeException: Received fatal alert: protocol_version

https://rustls.jbp.io/
javax.net.ssl.SSLHandshakeException: Received fatal alert: protocol_version

https://h2o.examp1e.net
h2
TLS_AES_128_GCM_SHA256
TLS_1_3
6209

https://mew.org/
h2
TLS_AES_128_GCM_SHA256
TLS_1_3
137

https://tls13.baishancloud.com/
javax.net.ssl.SSLHandshakeException: Received fatal alert: protocol_version

https://tls13.akamai.io/
java.net.ConnectException: Failed to connect to tls13.akamai.io/2a02:26f0:5c:1a3:0:0:0:5f37:443

https://swifttls.org/
http/1.0
TLS_AES_128_GCM_SHA256
TLS_1_3
231
```

**Comment by yschimke** [comment]
The alternative, leaving 1.3 in place and additionally adding in 1.3 CipherSuites carries some small risk as noted here.

> Negotiating with an older server
> 
>    Some legacy server implementations are known to not implement the TLS
>    specification properly and might abort connections upon encountering
>    TLS extensions or versions which they are not aware of.
>    Interoperability with buggy servers is a complex topic beyond the
>    scope of this document.  Multiple connection attempts may be required
>    in order to negotiate a backwards compatible connection; however,
>    this practice is vulnerable to downgrade attacks and is NOT
>    RECOMMENDED.

**Comment by yschimke** [comment]
```
Running tests using Jdk9Platform 11-ea+20
TLS1.3 only
https://enabled.tls13.com                TLS_1_3 TLS_AES_128_GCM_SHA256 h2 200 140b
https://www.howsmyssl.com/a/check        javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure
https://tls13.cloudflare.com             TLS_1_3 TLS_AES_128_GCM_SHA256 h2 521 4546b
https://www.allizom.org/robots.txt       TLS_1_3 TLS_AES_128_GCM_SHA256 h2 200 119b
https://tls13.crypto.mozilla.org/        TLS_1_3 TLS_AES_128_GCM_SHA256 http/1.1 200 3456b
https://tls.ctf.network/                 javax.net.ssl.SSLHandshakeException: Received fatal alert: protocol_version
https://rustls.jbp.io/                   javax.net.ssl.SSLHandshakeException: Received fatal alert: protocol_version
https://h2o.examp1e.net                  TLS_1_3 TLS_AES_128_GCM_SHA256 h2 200 6209b
https://mew.org/                         TLS_1_3 TLS_AES_128_GCM_SHA256 h2 200 137b
https://tls13.baishancloud.com/          javax.net.ssl.SSLHandshakeException: Received fatal alert: protocol_version
https://tls13.akamai.io/                 java.net.ConnectException: Failed to connect to tls13.akamai.io/2a02:26f0:5c:19f:0:0:0:5f37:443
https://swifttls.org/                    TLS_1_3 TLS_AES_128_GCM_SHA256 http/1.0 200 231b
TLS1.3 then fallback
https://enabled.tls13.com                TLS_1_3 TLS_AES_128_GCM_SHA256 h2 200 140b
https://www.howsmyssl.com/a/check        TLS_1_2 TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 http/1.1 200 468b
https://tls13.cloudflare.com             TLS_1_3 TLS_AES_128_GCM_SHA256 h2 521 4546b
https://www.allizom.org/robots.txt       TLS_1_3 TLS_AES_128_GCM_SHA256 h2 200 119b
https://tls13.crypto.mozilla.org/        TLS_1_3 TLS_AES_128_GCM_SHA256 http/1.1 200 3456b
https://tls.ctf.network/                 TLS_1_2 TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 http/1.1 200 263b
https://rustls.jbp.io/                   TLS_1_2 TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 http/1.1 200 416b
https://h2o.examp1e.net                  TLS_1_3 TLS_AES_128_GCM_SHA256 h2 200 6209b
https://mew.org/                         TLS_1_3 TLS_AES_128_GCM_SHA256 h2 200 137b
https://tls13.baishancloud.com/          TLS_1_2 TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 http/1.1 200 2737b
https://tls13.akamai.io/                 TLS_1_2 TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 http/1.1 200 4192b
https://swifttls.org/                    TLS_1_3 TLS_AES_128_GCM_SHA256 http/1.0 200 231b
TLS1.3+TLS1.2
https://enabled.tls13.com                TLS_1_3 TLS_AES_128_GCM_SHA256 h2 200 140b
https://www.howsmyssl.com/a/check        TLS_1_2 TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 http/1.1 200 518b
https://tls13.cloudflare.com             TLS_1_3 TLS_AES_128_GCM_SHA256 h2 521 4546b
https://www.allizom.org/robots.txt       TLS_1_3 TLS_AES_128_GCM_SHA256 h2 200 119b
https://tls13.crypto.mozilla.org/        TLS_1_3 TLS_AES_128_GCM_SHA256 http/1.1 200 3456b
https://tls.ctf.network/                 TLS_1_2 TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 http/1.1 200 263b
https://rustls.jbp.io/                   TLS_1_2 TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 http/1.1 200 416b
https://h2o.examp1e.net                  TLS_1_3 TLS_AES_128_GCM_SHA256 h2 200 6209b
https://mew.org/                         TLS_1_3 TLS_AES_128_GCM_SHA256 h2 200 137b
https://tls13.baishancloud.com/          TLS_1_2 TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 http/1.1 200 2837b
https://tls13.akamai.io/                 TLS_1_2 TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 http/1.1 200 4192b
https://swifttls.org/                    TLS_1_3 TLS_AES_128_GCM_SHA256 http/1.0 200 231b
```

File: okhttp/src/main/java/okhttp3/ConnectionSpec.java [hunk 7]
```diff
@@ -50,10 +50,20 @@
...
+      CipherSuite.TLS_CHACHA20_POLY1305_SHA256,
+      CipherSuite.TLS_AES_128_CCM_SHA256,
+      CipherSuite.TLS_AES_256_CCM_8_SHA256
+  };
+
   // This is nearly equal to the cipher suites supported in Chrome 51, current as of 2016-05-25.
   // All of these suites are available on Android 7.0; earlier releases support a subset of these
   // suites. https://github.com/square/okhttp/issues/1972
   private static final CipherSuite[] APPROVED_CIPHER_SUITES = new CipherSuite[] {
+
```
Comment (PromanSEW) [inline okhttp/src/main/java/okhttp3/ConnectionSpec.java]: Extra line?

File: okhttp/src/main/java/okhttp3/CipherSuite.java [hunk 8]
```diff
@@ -389,12 +389,19 @@
   // public static final CipherSuite TLS_ECDHE_ECDSA_WITH_AES_256_CCM_8 = of("TLS_ECDHE_ECDSA_WITH_AES_256_CCM_8", 0xc0af);
   public static final CipherSuite TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256 = of("TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256", 0xcca8);
   public static final CipherSuite TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256 = of("TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256", 0xcca9);
-  // public static final CipherSuite TLS_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256 = of("TLS_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256", 0xccaa);
+  public static final CipherSuite TLS_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256 = of("TLS_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256", 0xccaa);
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/CipherSuite.java]: FYI: https://bugs.openjdk.java.net/browse/JDK-8204192

**Comment by yschimke** [comment]
Would like to merge this for testing in 3.12-SNAPSHOT

**Pull request merged**